### PR TITLE
Bumps deps

### PR DESCRIPTION
**Pull request by jmarrec**
Companion PR: https://github.com/NREL/OpenStudio/pull/4596

Latest run: https://github.com/jmarrec/conan-openstudio-ruby/runs/6564537150?check_suite_focus=true

----

Original Post:

It's running on my fork, https://github.com/jmarrec/conan-openstudio-ruby/runs/5785999801?check_suite_focus=true

I have an issue...:
* openssl 1.1.1n won't build on GCC 7 and 8 on CI (it works fine locally on my ubuntu 20.04 machine...)
```shell
 /home/conan/.conan/data/openssl/1.1.1n/_/_/package/d50a0d523d98c15bb147b18fa7d203887c38be8b/lib/libcrypto.a(threads_pthread.o): In function `fork_once_func':
threads_pthread.c:(.text+0x16): undefined reference to `pthread_atfork'
collect2: error: ld returned 1 exit status
```

* openssl 1.1.0l won't build **in debug** on msvc-2022 (and there is no binary package for it):
```
2022-04-01T10:39:08.7868654Z 	set ASM=nasm
2022-04-01T10:39:08.7869883Z 	"C:\.conan\18ba5c\1\bin\perl.exe" "crypto\aes\asm\aes-x86_64.pl" auto crypto\aes\aes-x86_64.asm
2022-04-01T10:39:09.0013273Z 	nasm -f win64 -DNEAR -Ox -g -ocrypto\aes\aes-x86_64.obj "crypto\aes\aes-x86_64.asm"
2022-04-01T10:39:09.8242544Z NMAKE : fatal error U1077: 'C:\Users\runneradmin\.conan\data\nasm\2.14\_\_\package\01edd76db8e16db9b38c3cca44ec466a9444c388\bin\nasm.EXE' : return code '0xc0000005'
2022-04-01T10:39:09.8243466Z Stop.
2022-04-01T10:39:09.8274678Z NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.31.31103\bin\HostX64\x64\nmake.exe"' : return code '0x2'
2022-04-01T10:39:09.8275272Z Stop.
```

### CODE REVIEW

**Comment by jmarrec** [comment]
Holy crap, the linking command is so superbly wrong... I'm always amazed by the fact that ruby build system actually works in some cases...

```
ORIGINAL LINK FLAGS = -Wl,-rpath,/home/conan/.conan/data/openstudio_ruby/2.7.2/jmarrec/testing/build/9f9bc9c7ba86f123c671ad0cd57bd027161f8634/Ruby-prefix/src/Ruby-install/lib
 -lruby-static -lz -lpthread -lrt -lrt -lgmp -ldl -lcrypt -lm -lz -lpthread
 -lrt -lrt -lgmp -ldl -lcrypt -lm -lm -lanl -lssl -lcrypto
 -lreadline -lncurses -lgdbm_compat -lgdbm -lz -ldl -lffi -lutil -lyaml -lanl -lssl -lcrypto 
 -lreadline -lncurses -lgdbm_compat -lgdbm -lz -ldl -lffi -lutil -lyaml
```

Here are the link flags that work with conanio/gcc7: Removing duplicates, moving `-pthread` (not `-lpthread` at the end)
```
NEW LINK FLAGS =  -Wl,-rpath,/home/conan/.conan/data/openstudio_ruby/2.7.2/jmarrec/testing/build/9f9bc9c7ba86f123c671ad0cd57bd027161f8634/Ruby-prefix/src/Ruby-install/lib
 -lruby-static -lz -lrt -lgmp -ldl -lcrypt -lm -lanl -lssl -lcrypto 
 -lreadline -lncurses -lgdbm_compat -lgdbm -lffi -lutil -lyaml -pthread
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -222,14 +222,33 @@ if( UNIX )
   # brew install autoconf automake libtool
   # brew link autoconf automake libtool
 
+  if(CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0)
+    set(patchCommand PATCH_COMMAND pwd
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/Ruby.patch
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/Ruby.util.patch
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0002-patch-encoding-CP720.patch
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0003-YYUSE.patch ${EXTRA_PATCH}
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0004-GCC7_8_fix_link_flags.patch
```
Comment (jmarrec) [inline CMakeLists.txt:225]: New patch for GCC 7 and 8.... it sucks to have to do yet another weird config trick, but that's the way it is

File: CMakeLists.txt [hunk 2]
```diff
@@ -260,10 +279,10 @@ else()
   #file(DOWNLOAD https://iweb.dl.sourceforge.net/project/gnuwin32/bison/2.4.1/bison-2.4.1-bin.zip ${CMAKE_BINARY_DIR}/tools/bison-2.4.1-bin.zip)
   #file(DOWNLOAD https://iweb.dl.sourceforge.net/project/gnuwin32/bison/2.4.1/bison-2.4.1-dep.zip ${CMAKE_BINARY_DIR}/tools/bison-2.4.1-dep.zip)
   if(NOT EXISTS ${CMAKE_BINARY_DIR}/tools/sed-4.2.1-dep.zip)
-    file(DOWNLOAD https://iweb.dl.sourceforge.net/project/gnuwin32/sed/4.2.1/sed-4.2.1-dep.zip ${CMAKE_BINARY_DIR}/tools/sed-4.2.1-dep.zip)
+    file(DOWNLOAD  https://sourceforge.net/projects/gnuwin32/files/sed/4.2.1/sed-4.2.1-dep.zip ${CMAKE_BINARY_DIR}/tools/sed-4.2.1-dep.zip)
```
Comment (jmarrec) [inline CMakeLists.txt]: sourceforge endpoint changed

File: conanfile.py [hunk 3]
```diff
@@ -84,28 +84,24 @@ def requirements(self):
         """
         Declare required dependencies
         """
-        self.requires("openssl/1.1.0l") # fails with 1.1.1h https://github.com/openssl/openssl/issues/3884`
+        # Doesn't work with 3.x.
+        # Doesn't work on gcc 7 and 8 with 1.1.1n
+        self.requires("openssl/1.1.1n")
```
Comment (jmarrec) [inline conanfile.py]: bump all deps

File: patches/0003-YYUSE.patch [hunk 4]
```diff
@@ -0,0 +1,12 @@
+diff --git a/parse.y b/parse.y
+index 0fd0c46..97c9743 100644
+--- a/parse.y
++++ b/parse.y
+@@ -12673,7 +12673,6 @@ count_char(const char *str, int c)
+ RUBY_FUNC_EXPORTED size_t
+ rb_yytnamerr(struct parser_params *p, char *yyres, const char *yystr)
+ {
+-    YYUSE(p);
```
Comment (jmarrec) [inline patches/0003-YYUSE.patch:9]: Patch to Remove YYUSE (which is now private in bison 3.7.5+)

cf https://github.com/ruby/ruby/commit/b091889ed67f278b1652ddde88596e2160d5aef0#diff-60e6f0aa6e190c54f6cb0ac74148ced3cf0ef8703e1c3824a07a82dc54654953R13335

File: CMakeLists.txt [hunk 5]
```diff
@@ -222,14 +222,33 @@ if( UNIX )
   # brew install autoconf automake libtool
   # brew link autoconf automake libtool
 
+  if(CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0)
+    set(patchCommand PATCH_COMMAND pwd
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/Ruby.patch
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/Ruby.util.patch
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0002-patch-encoding-CP720.patch
+      && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0003-YYUSE.patch ${EXTRA_PATCH}
```
Comment (jmarrec) [inline CMakeLists.txt:224]: new patch for all...

File: CMakeLists.txt [hunk 6]
```diff
@@ -301,6 +321,7 @@ else()
                && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/Ruby.win.272.patch  # && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/Ruby.nodynamic.patch
                && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0001-patch-to-support-version-ranges-for-MSVC.patch
                && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0002-patch-encoding-CP720.patch
+               && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/patches/0003-YYUSE.patch
```
Comment (jmarrec) [inline CMakeLists.txt:318]: including windows

File: patches/0004-GCC7_8_fix_link_flags.patch [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
++		$(ECHO) EXTOBJS = $(EXTOBJS)
++		$(ECHO) LIBRUBYARG = $(LIBRUBYARG)
++		$(ECHO) MAINLIBS = $(MAINLIBS)
++		$(ECHO) LIBS = $(LIBS)
++		$(ECHO) EXTLIBS = $(EXTLIBS)
++		$(ECHO) OUTFLAG = $(OUTFLAG)
++		$(ECHO) POSTLINK = $(POSTLINK)
++		$(ECHO) ORIGINAL LINK FLAGS =  $(LIBRUBYARG) $(MAINLIBS) $(LIBS) $(EXTLIBS)
++		$(ECHO) NEW LINK FLAGS = -Wl,-rpath,$(libdir) -lruby-static -lz -lrt -lgmp -ldl -lcrypt -lm -lanl -lssl -lcrypto -lreadline -lncurses -lgdbm_compat -lgdbm -lffi -lutil -lyaml -pthread
++		$(Q) $(PURIFY) $(CC) $(LDFLAGS) $(XLDFLAGS) $(MAINOBJ) $(EXTOBJS) -Wl,-rpath,$(libdir) -lruby-static -lz -lrt -lgmp -ldl -lcrypt -lm -lanl -lssl -lcrypto -lreadline -lncurses -lgdbm_compat -lgdbm -lffi -lutil -lyaml -pthread $(OUTFLAG)$@
```
Comment (jmarrec) [inline patches/0004-GCC7_8_fix_link_flags.patch:25]: Patch for the link flags...

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -128,23 +128,23 @@ jobs:
           set -x
 
           echo "=========== MAKEFILE ================"
-          find ~/.conan/data/openstudio_ruby/2.7.2/nrel/testing/build -name "Makefile"|while read fname; do
+          find ~/.conan/data/openstudio_ruby/2.7.2/$CONAN_USERNAME/testing/build -name "Makefile"|while read fname; do
```
Comment (jmarrec) [inline .github/workflows/build.yml:121]: make debug work on forks too

**Comment by jmarrec** [comment]
Latest run on my fork: https://github.com/jmarrec/conan-openstudio-ruby/runs/6563992930?check_suite_focus=true

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -20,7 +20,7 @@ jobs:
       # fail-fast: Default is true, switch to false to allow one platform to fail and still run others
       fail-fast: false
       matrix:
-        build_name: [GCC-7, GCC-8, GCC-9, GCC-10, GCC-11, Apple-Clang 11, Apple-Clang 12, MSVC-2022, MSVC-2019, MSVC-2017]
+        build_name: [GCC-7, GCC-8, GCC-9, GCC-10, GCC-11, Apple-Clang 11, Apple-Clang 12, MSVC-2022, MSVC-2019]
```
Comment (jmarrec) [inline .github/workflows/build.yml:23]: dropping MSVC 2017 (this will be required on OS side if we want to support 2022 anyways)

File: build.py [hunk 10]
```diff
@@ -52,18 +49,22 @@
...
-    #                             "ruby_installer:build_type": "Release",
-    #                         })
+    builder.update_build_if(
+        lambda build: True,
+        new_settings={
+            "bison:build_type": "Release",
+            "ruby_installer:build_type": "Release",
+            "m4:build_type": "Release",
+        },
+    )
```
Comment (jmarrec) [inline build.py:59]: I had tried to pass a  specific build profile to the docker but failed, cf revert in cd54db977ffb27534343f7434e7823e8e9067540

**Comment by jmarrec** [comment]
Ok this is ready to go: https://github.com/jmarrec/conan-openstudio-ruby/runs/6564537150?check_suite_focus=true

I have tested locally that I can build OpenStudio via https://github.com/NREL/OpenStudio/pull/4596

File: .github/workflows/build.yml (hunk 9 above)
Comment (tijcolem) [reply .github/workflows/build.yml:23]: Good with me. We need to move to 2022 soon

**Review by tijcolem — APPROVED** [review]

**Pull request merged**